src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,296 @@
+    obtain ⟨z1, z2, hz1, hz2, hz3⟩ := hy1,
+    use z1, use z2, split, rw set.mem_image, use z1, refine ⟨hz1, rfl⟩,
+    split, rw set.mem_image, use z2, refine ⟨hz2, rfl⟩, tidy, }
+end
+
+lemma is_homogeneous_ideal.sup {I J : ideal R}
+  (HI : is_homogeneous_ideal A I) (HJ : is_homogeneous_ideal A J) :
+  is_homogeneous_ideal A (I ⊔ J) :=
+begin
+  rw is_homogeneous_ideal_iff_exists at HI HJ ⊢,
+  obtain ⟨⟨s₁, rfl⟩, ⟨s₂, rfl⟩⟩ := ⟨HI, HJ⟩,
+  refine ⟨s₁ ∪ s₂, _⟩,
+  rw [set.image_union],
+  exact (submodule.span_union _ _).symm,
+end
+
+lemma is_homogeneous_ideal.Sup {ℐ : set (ideal R)} (Hℐ : ∀ (I ∈ ℐ), is_homogeneous_ideal A I) :
+  is_homogeneous_ideal A (Sup ℐ) :=
+begin
+  simp_rw [is_homogeneous_ideal_iff_exists] at Hℐ,
+  set 𝓈 : ℐ → set (homogeneous_submonoid A) := λ I : ℐ, Exists.some (Hℐ I _) with 𝓈_eq,
+  have h𝓈 : ∀ I : ℐ, I.1 = ideal.span (coe '' 𝓈 I) := λ I : ℐ, Exists.some_spec (Hℐ I _),
+  rw is_homogeneous_ideal_iff_exists,
+  use sUnion (set.range 𝓈),
+  rw [sUnion_range, image_Union, ideal.span, submodule.span_Union],
+  ext r, split,
+  { suffices : Sup ℐ ≤ _, revert r, exact this,
+    rw Sup_le_iff, intros I HI,
+    have ineq1 : I ≤ ideal.span I := ideal.subset_span, refine le_trans ineq1 _,
+    rw ideal.span_le, simp only [coe_subset_coe, ideal.submodule_span_eq], intros x hx,
+    rw submodule.mem_supr, intros J HJ,
+    apply HJ ⟨I, HI⟩, rw ←h𝓈 ⟨I, HI⟩, assumption },
+  { suffices : _ ≤  Sup ℐ, revert r, exact this,
+    rw supr_le_iff, intros I, rw submodule.span_le, intros x hx,
+    simp only [mem_image] at hx, obtain ⟨x', hx1, hx2⟩ := hx,
+    simp only [mem_coe, subtype.val_eq_coe], dsimp only at hx1,
+    apply ideal.mem_Sup_of_mem, use I.2,
+    simp only [subtype.val_eq_coe] at h𝓈 ⊢, rw h𝓈,
+    refine ideal.subset_span _, rw [mem_image], use x', refine ⟨hx1, hx2⟩, },
+  intros I, exact I.2,
+end
+
+variables (A)
+
+instance : has_inf (homogeneous_ideal A) :=
+{ inf := λ I J, ⟨I ⊓ J, I.prop.inf J.prop⟩ }
+
+instance : has_Inf (homogeneous_ideal A) :=
+{ Inf := λ ℐ, ⟨Inf (coe '' ℐ), homogeneous_ideal.Inf $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+instance : has_sup (homogeneous_ideal A) :=
+{ sup := λ I J, ⟨I ⊔ J, I.prop.sup J.prop⟩ }
+
+instance : has_Sup (homogeneous_ideal A) :=
+{ Sup := λ ℐ, ⟨Sup (coe '' ℐ), is_homogeneous_ideal.Sup $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+instance : has_mul (homogeneous_ideal A) :=
+{ mul := λ I J, ⟨I * J, I.prop.mul J.prop⟩ }
+
+instance : has_add (homogeneous_ideal A) := ⟨(⊔)⟩
+
+end operations
+
+section of_ideal
+
+open set_like direct_sum set
+
+variables {ι R : Type*} [comm_ring R]
+variables (A : ι → submodule R R)
+variable (I : ideal R)
+
+/--For any `I : ideal R`, not necessarily homogeneous, there is a homogeneous ideal asscoiated with
+`I` spaned by all homogeneous elements in `I`. This construction will be used when proving radical
+of homogeneous ideal is homogeneous-/
+def homogeneous_ideal_of_ideal : ideal R := ideal.span (set_of (is_homogeneous A) ∩ I)
```

Comment:
I suggest calling this `ideal.homogenize`.

Reply:
Presumably this function is monotone, if not also a Galois connection?

Reply:
Correct me if I am wrong, but I think this is not a galois connection. Consider ideal generated by x+y in C[x,y], then it's homogenization is trivial, thus subset of any ideal.

Reply:
Ah sorry, I thought `homogenization` made the ideal bigger not smaller, like `span`.

Reply:
that is what i worry about name homogenisation, it seems suggesting making ideal bigger.

Reply:
Is it true that `I.homogenization` is `Sup { x | x.is_homogeneous ∧ x ≤ I }`

Reply:
yes it is. I don't know if this is going to be useful, but I include a proof
```lean
lemma ideal.homogenisation.eq_Sup [Π (i : ι) (x : A i), decidable (x ≠ 0)] :
  ideal.homogenisation A I = Sup { J : ideal R | ideal.is_homogeneous A J ∧ J ≤ I }
```
at line 302.

Reply:
Though I don't think this construction is to be useful outside proving radical of hom ideal is hom.

Reply:
Should we have both ways of homogenizing? Maybe the current construction is `homogeneous_core` and the other is `homogeneous_hull`?

Reply:
I have added `homogenous_hull`